frontend/controller/serviceworkers/sw-primary.js

```diff
@@ -83,36 +223,73 @@ self.addEventListener('message', function (event) {
+        // order is visible, prerender, hidden
+          if (a.visibilityState === 'visible') return -1
+          if (b.visibilityState === 'visible') return 1
+          if (a.visibilityState === 'hidden') return 1
+          if (b.visibilityState === 'hidden') return -1
+        }
 
-  if (!(self.Notification && self.Notification.permission === 'granted')) {
-    console.debug("[sw] received a push notification but aren't displaying it due to the permission not granted")
-    return
-  }
-
-  const data = event.data.json()
-  console.debug('[sw] push received: ', data)
+        return 0
+      })
+      if (!clientList.length) {
+        return self.clients.openWindow(`/app${event.notification.data.path ?? '/'}`)
+      }
+      const client = clientList[0]
+      if (event.notification.data?.path) {
+        client.postMessage({
+          type: 'navigate',
+          path: event.notification.data.path
+        })
+      }
+      if (!client.focused) return client.focus()
+    }))
+})
 
-  self.registration.showNotification(
-    data.title,
-    {
-      body: data.body || '',
-      icon: '/assets/images/pwa-icons/group-income-icon-transparent.svg'
+sbp('okTurtles.events/on', KV_EVENT, ({ contractID, key, data }) => {
+  const rootState = sbp('chelonia/rootState')
+  switch (key) {
+    case KV_KEYS.LAST_LOGGED_IN: {
+      if (!rootState.lastLoggedIn) rootState.lastLoggedIn = Object.create(null)
+      rootState.lastLoggedIn[contractID] = data
+      break
+    }
+    case KV_KEYS.UNREAD_MESSAGES: {
+      if (!rootState.chatroom) rootState.chatroom = Object.create(null)
+      rootState.chatroom.unreadMessages = data
+      break
+    }
+    case KV_KEYS.PREFERENCES: {
+      rootState.preferences = data
+      break
     }
-  )
+    case KV_KEYS.NOTIFICATIONS: {
+      if (!rootState.notifications) rootState.notifications = Object.create(null)
+      rootState.notifications.status = data
+      break
+    }
+  }
 })
 
-self.addEventListener('pushsubscriptionchange', async function (event) {
-  // NOTE: Currently there is no specific way to validate if a push-subscription is valid. So it has to be handled in the front-end.
-  // (reference:https://pushpad.xyz/blog/web-push-how-to-check-if-a-push-endpoint-is-still-valid)
-  const subscription = await self.registration.pushManger.subscribe(event.oldSubscription.options)
+sbp('okTurtles.events/on', NEW_CHATROOM_UNREAD_POSITION, ({ chatRoomID, messageHash }) => {
+  const rootState = sbp('chelonia/rootState')
+  if (messageHash) {
+    if (!rootState.chatroom) rootState.chatroom = Object.create(null)
+    if (!rootState.chatroom.chatRoomScrollPosition) rootState.chatroom.chatRoomScrollPosition = Object.create(null)
 
-  // Sending the client a message letting it know of the subscription change.
-  await sendMessageToClient({
-    type: 'pushsubscriptionchange',
-    subscription
-  })
+    rootState.chatroom.chatRoomScrollPosition[chatRoomID] = messageHash
```

Comment:
How/where does this update get saved to IndexedDB?

Reply:
They don't. Updated.